Allow for explain to be set on a query

kyle-mccarthy opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | n/a

Allow for explain to be set on a query, so you can see how the score for a certain document is calculated.

https://www.elastic.co/guide/en/elasticsearch/reference/current/search-explain.html

Can be used by `ElasticActiveRecord::find()->query(...)->explain(true)->all();`.

cebe commented on ActiveRecord.php:
```diff
@@ -193,6 +193,14 @@ public function getHighlight()
 
     /**
+     * @return array|null An explanation for each hit on how its score was computed
+     */
```
need `@since 2.0.5` annotation.

cebe commented on Query.php:
```diff
@@ -171,6 +171,11 @@ class Query extends Component implements QueryInterface
+    /**
+     * @var Enables explanation for each hit on how its score was computed.
+     * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-explain.html
+     */
```
need `@since 2.0.5` annotation.

cebe commented on Query.php:
```diff
@@ -688,4 +693,16 @@ public function postFilter($filter)
+     * @param $explain
+     * @return $this
+     * @see $explain
+     */
```
need `@since 2.0.5` annotation.

kyle-mccarthy commented:
@cebe Okay I have updated the annotations and added in two simple test cases to check for the existence of `_explanation` on the query when it is set and to ensure it doesn't exist when not set.  Let me know if you see anything else.

cebe commented:
looks good to me, thanks! If you add a CHANGLOG line I'll merge it right away.

cebe commented on Query.php:
```diff
@@ -171,6 +171,12 @@ class Query extends Component implements QueryInterface
      */
     public $options = [];
+    /**
+     * @var Enables explanation for each hit on how its score was computed.
```
oh, missed that. this should include the type: `@var bool`

cebe commented on Query.php:
```diff
@@ -688,4 +694,17 @@ public function postFilter($filter)
+
+    /**
+     * Explain for how the score of each document was computer
+     * @param $explain
```
same here: `@param bool $explain`

kyle-mccarthy commented:
Done and not a problem!

cebe commented:
Merged, thank you!
